src/main/java/net/fabricmc/loader/impl/util/version/VersionBoundsImpl.java

```diff
@@ -0,0 +1,393 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.loader.impl.util.version;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Objects;
+
+import net.fabricmc.loader.api.SemanticVersion;
+import net.fabricmc.loader.api.Version;
+import net.fabricmc.loader.api.metadata.version.VersionBounds;
+
+public final class VersionBoundsImpl implements VersionBounds {
+	static final VersionBounds NONE = new VersionBoundsImpl(null, false, null, false);
+
+	private final Version min;
+	private final boolean minInclusive;
+	private final Version max;
+	private final boolean maxInclusive;
+
+	VersionBoundsImpl(Version min, boolean minInclusive,
+			Version max, boolean maxInclusive) {
+		this.min = min;
+		this.minInclusive = min != null ? minInclusive : false;
+		this.max = max;
+		this.maxInclusive = max != null ? maxInclusive : false;
+
+		assert min != null || !minInclusive;
+		assert max != null || !maxInclusive;
+		assert min == null || min instanceof SemanticVersion || minInclusive;
+		assert max == null || max instanceof SemanticVersion || maxInclusive;
+		assert min == null || max == null || min instanceof SemanticVersion && max instanceof SemanticVersion || min.equals(max);
+	}
+
+	@Override
+	public boolean isSemantic() {
+		return (min == null || min instanceof SemanticVersion)
+				&& (max == null || max instanceof SemanticVersion);
+	}
+
+	@Override
+	public Version getMin() {
+		return min;
+	}
+
+	@Override
+	public boolean isMinInclusive() {
+		return minInclusive;
+	}
+
+	@Override
+	public Version getMax() {
+		return max;
+	}
+
+	@Override
+	public boolean isMaxInclusive() {
+		return maxInclusive;
+	}
+
+	@Override
+	public boolean equals(Object obj) {
+		if (obj instanceof VersionBounds) {
+			VersionBounds o = (VersionBounds) obj;
+
+			return Objects.equals(min, o.getMin()) && minInclusive == o.isMinInclusive()
+					&& Objects.equals(max, o.getMax()) && maxInclusive == o.isMaxInclusive();
+		} else {
+			return false;
+		}
+	}
+
+	@Override
+	public String toString() {
+		if (min == null) {
+			if (max == null) {
+				return "(-∞,∞)";
+			} else {
+				return String.format("(-∞,%s%c", max, maxInclusive ? ']' : ')');
+			}
+		} else if (max == null) {
+			return String.format("%c%s,∞)", minInclusive ? '[' : '(', min);
+		} else {
+			return String.format("%c%s,%s%c", minInclusive ? '[' : '(', min, max, maxInclusive ? ']' : ')');
+		}
+	}
```

Comment:
I don't think it's a good idea to use mathematical notation for a `toString` implementation. I'd avoid using `toString` for user-readable information, and even then, this notation in general is not user-friendly.

Reply:
This isn't supposed to be interpreted by the user either, IMO it does a fair job for its main applications (devs working on version analysis similar to my installable mod inference code).

Reply:
I would say `+inf` is better than `+∞`, just in case some people's computer is not UTF-8 encoding.

Reply:
Java should convert it properly, I'd rather have a more compact format for this niche thing than not.